rfc/expr-context.md

```diff
@@ -0,0 +1,115 @@
+# Expression Context
+
+Cosmo Router provides a context object to an expr-lang VM that can be used to access information about the request, response, and other relevant information. The context object is available at all times during the request's lifecycle.
+
+## request
+
+The client request object is a read-only object that provides information about the incoming request. It is available at all times during the request's lifecycle.
+
+```
+request.id
+request.header
+request.method
+request.body.query
+request.body.operation_name
+request.body.variables
+request.body.extensions
+request.uri.host
+request.uri.path
+request.uri.port
+
+request.client.name
+request.client.version
+request.client.ip
+request.routerConfigVersion
+```
+
+# request.context
+
+The context object provides information about the context of the request. A generic object that can be used to store any information that is relevant to the request.
+
+```
+request.context.agent
+```
+
+In the Router config, you can use the following expression to set a custom field in the context object:
+
+```yaml
+context:
+  agent: "request.auth.claims.sub || request.client.ip || request.header.X-Forwarded-For"
+```
+
+At runtime, the following expression can be used to set additional fields in the context object or to override or unset existing fields:
+
+```
+# set agent to undefined
+unset("request.context.agent")
```

Comment:
Expr is side effect free, this won't be possible.

Reply:
You can use a function:
https://expr-lang.org/docs/functions

Reply:
It is a principle. We should avoid that. Expr should only be used to get / compute a value.